DataBiosphere/Saturn-UI
# [AN-308] Enable browsing genomic data with DRS URIs in IGV

### Jira ticket: https://broadworkbench.atlassian.net/browse/AN-308




## Summary of changes
This lets users search and visualize VCFs, BAMs, CRAMs, BEDs etc. through DRS URIs in the Terra IGV app.

Here's how it looks!

https://github.com/user-attachments/assets/b33637c8-a039-4c38-a0a8-fc3776998067

In brief, upon the user choosing to open IGV for a selected data table row, Terra UI now finds any DRS URIs among the row values, [resolves](https://drshub.dsde-prod.broadinstitute.org/#/drsHub/resolveDrs) those DRS URIs to eligible genomic file extensions and corresponding index files, and if found resolves access URLs (a.k.a. signed URLs) to the underlying files, then sets those as IGV tracks as currently done for files behind HTTPS or GS URLs.  

That enables VCFs, BAMs, and CRAMs from data repositories like AnVIL to leverage genome browsing in Terra.  Incidentally, this also enables seeing genomic features upon searching most genes without needing to zoom several times.

### Background
DRS is a GA4GH standard that addresses a cloud resource independently from the particular cloud's infrastructure.  [DRS URIs have been supported in Terra](https://support.terra.bio/hc/en-us/articles/360039330211-Overview-Interoperable-data-GA4GH-DRS-URIs) for some time.  This extends that support to the Terra IGV app.  

The enhancement is available as a preview behind a new "IGV enhancements" feature flag that corresponds to the "[IGV is easier to find and use with data in a workspace](https://support.terra.bio/hc/en-us/articles/31191428549659-IGV-is-easier-to-find-and-use-with-data-in-a-workspace)" item in the [Terra roadmap](https://support.terra.bio/hc/en-us/sections/30968105851931-Terra-Roadmap).

This was implemented in support of the AnVIL project, which stores data and makes it available via DRS URIs in requester-pays buckets.  It enables public and controlled-access data in AnVIL to be browsed in genomic context, after it's been exported to Terra.  Assuming the underlying buckets have a suitable CORS configuration, this enhancement also benefits other data repositories that support DRS and export eligible data to Terra.  

The sequence diagram below conveys the dataflow.

![IGV_for_DRS_files_in_Terra UI__CORS_configuration_sequence_diagram_2024-10-28](https://github.com/user-attachments/assets/91f76424-5911-4b86-a770-d7a2dddee858)

Related [discussion in #dsp-infosec-champions](https://broadinstitute.slack.com/archives/CADU7L0SZ/p1729520517663919) has more security and compliance context.

### Limitations
#### Requires genomic file and index file on same row
In this first phase of support, DRS URIs for the genomic file and its index file must be on the same row of a data table.  Much (most?) data exported to Terra from e.g. AnVIL has such file pairs split _across_ rows.  [AN-310](https://broadworkbench.atlassian.net/browse/AN-310) tracks making IGV usable for that common case.

#### Access to genomic data can expire before Terra session
Access URLs that get resolved from DRS URIs often expire before the user’s Terra session expires.  This could cause in-depth genomic browsing to unexpectedly halt for the user.  Refreshing etc. can work around the issue, but a more robust solution would be better.  This is tracked in [AN-309](https://broadworkbench.atlassian.net/browse/AN-309).

### Testing strategy

Automated tests verify new behavior.  To manually test:

1.  Go to https://pr-10-dot-bvdp-saturn-dev.appspot.com/#feature-preview
2.  Enable "Enhancements to IGV genome browser integration" feature preview
3.  Go to https://pr-10-dot-bvdp-saturn-dev.appspot.com/#workspaces/anvil-dev-developer-work/igv_drs_test_area/data.
4.  Select checkbox for 1st row in that file_inventory table.
5.  Click "Open With..."
6.  Click "IGV"
7.  Confirm "Searching for valid files with indices..." briefly appears
8.  Confirm "NA06985.haplotypeCalls.er.raw.g.vcf.gz" appears after a few seconds
9.  Select that file, click "Launch IGV"
10.  Confirm IGV loads, and that you see a track labeled "NA06985.haplotypeCalls.er.ra..."
11.  In the IGV search box, type in "LDLR", then press "Enter" / "return" on your keyboard
12.  Confirm you see faint blue and grey features in the NA06985 track


[AN-310]: https://broadworkbench.atlassian.net/browse/AN-310?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
